lib/services/local/presence.go

```diff
@@ -678,72 +678,138 @@ func (s *PresenceService) DeleteAllTunnelConnections() error {
 	}
-	return nil
+	rc.SetRevision(lease.Revision)
+	return rc, nil
 }
 
 // UpdateRemoteCluster updates selected remote cluster fields: expiry and labels
 // other changed fields will be ignored by the method
-func (s *PresenceService) UpdateRemoteCluster(ctx context.Context, rc types.RemoteCluster) error {
+func (s *PresenceService) UpdateRemoteCluster(ctx context.Context, rc types.RemoteCluster) (types.RemoteCluster, error) {
 	if err := services.CheckAndSetDefaults(rc); err != nil {
-		return trace.Wrap(err)
-	}
-	existingItem, update, err := s.getRemoteCluster(ctx, rc.GetName())
-	if err != nil {
-		return trace.Wrap(err)
+		return nil, trace.Wrap(err)
 	}
-	update.SetExpiry(rc.Expiry())
-	update.SetLastHeartbeat(rc.GetLastHeartbeat())
-	update.SetConnectionStatus(rc.GetConnectionStatus())
-	update.SetMetadata(rc.GetMetadata())
 
-	rev := update.GetRevision()
-	updateValue, err := services.MarshalRemoteCluster(update)
-	if err != nil {
-		return trace.Wrap(err)
-	}
-	updateItem := backend.Item{
-		Key:      backend.Key(remoteClustersPrefix, update.GetName()),
-		Value:    updateValue,
-		Expires:  update.Expiry(),
-		Revision: rev,
-	}
+	// Small retry loop to catch cases where there's a concurrent update which
+	// could cause conditional update to fail. This is needed because of the
+	// unusual way updates are handled in this method meaning that the revision
+	// in the provided remote cluster is not used. We should eventually make a
+	// breaking change to this behavior.
+	const iterationLimit = 3
+	for i := 0; i < iterationLimit; i++ {
+		existing, err := s.GetRemoteCluster(ctx, rc.GetName())
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		existing.SetExpiry(rc.Expiry())
+		existing.SetLastHeartbeat(rc.GetLastHeartbeat())
+		existing.SetConnectionStatus(rc.GetConnectionStatus())
+		existing.SetMetadata(rc.GetMetadata())
 
-	_, err = s.CompareAndSwap(ctx, *existingItem, updateItem)
-	if err != nil {
-		if trace.IsCompareFailed(err) {
-			return trace.CompareFailed("remote cluster %v has been updated by another client, try again", rc.GetName())
+		updateValue, err := services.MarshalRemoteCluster(existing)
+		if err != nil {
+			return nil, trace.Wrap(err)
 		}
-		return trace.Wrap(err)
+
+		lease, err := s.ConditionalUpdate(ctx, backend.Item{
+			Key:      backend.Key(remoteClustersPrefix, existing.GetName()),
+			Value:    updateValue,
+			Expires:  existing.Expiry(),
+			Revision: existing.GetRevision(),
+		})
+		if err != nil {
+			if trace.IsCompareFailed(err) {
+				// Retry!
+				continue
+			}
+			return nil, trace.Wrap(err)
+		}
+		existing.SetRevision(lease.Revision)
+		return existing, nil
+	}
+	return nil, trace.CompareFailed("failed to update remote cluster within %v iterations", iterationLimit)
+}
+
+// GetAndUpdateRemoteCluster fetches a remote cluster and then calls updateFn
+// to apply any changes, before persisting the updated remote cluster.
+func (s *PresenceService) GetAndUpdateRemoteCluster(
```

Comment:
This method is introduced to provide a foundation for a gRPC UpdateRemoteCluster which will support field masking, similar to the implementation for DeviceTrust.